cstrml/payment/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(option_result_contains)]
+
+use codec::{Decode, Encode, HasCompact};
+use frame_support::{
+    decl_event, decl_module, decl_storage, decl_error, dispatch::DispatchResult, Parameter,
+    traits::
+    {
+        Randomness, schedule::Named as ScheduleNamed,
+        Currency, ReservableCurrency
+    }
+};
+use sp_std::{prelude::*, convert::{TryInto}};
+use system::{ensure_root};
+use sp_runtime::{traits::{StaticLookup, Dispatchable, Zero, Convert}};
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+// Crust runtime modules
+use primitives::{
+    constants::time::MINUTES
+};
+
+use market::{OrderStatus, MarketInterface, Payment};
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub type BalanceOf<T> =
+    <<T as Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+impl<T: Trait> Payment<<T as system::Trait>::AccountId,
+    <T as system::Trait>::Hash, BalanceOf<T>> for Module<T>
+{
+    fn pay_sorder(client: &<T as system::Trait>::AccountId,
+                  provider: &<T as system::Trait>::AccountId,
+                  value: BalanceOf<T>) -> T::Hash {
+        let bn = <system::Module<T>>::block_number();
+        let bh: T::Hash = <system::Module<T>>::block_hash(bn);
+        let seed = [
+            &bh.as_ref()[..],
+            &client.encode()[..],
+            &provider.encode()[..],
+        ].concat();
+
+        // it can cover most cases, for the "real" random
+        let sorder_id = T::Randomness::random(seed.as_slice());
+        if T::Currency::reserve(&client, value).is_ok() {
+            <Payments<T>>::insert(sorder_id, PaymentLedger{
+                total: value,
+                already_paid: Zero::zero(),
+                already_unreserved: Zero::zero()
+            });
+        }
+        return sorder_id;
+    }
+
+    fn start_delayed_pay(sorder_id: &T::Hash) {
+        let sorder = T::MarketInterface::maybe_get_sorder(sorder_id).unwrap_or_default();
+        let interval =  TryInto::<T::BlockNumber>::try_into(MINUTES).ok().unwrap();
+        let times = (sorder.expired_on - sorder.completed_on)/MINUTES + 1;
+        let total = Self::payments(sorder_id).unwrap_or_default().total;
+        let piece_value: BalanceOf<T> = <T::CurrencyToBalance as Convert<u128, BalanceOf<T>>>::convert(<T::CurrencyToBalance as Convert<BalanceOf<T>, u128>>::convert(total)/times as u128 + 1);
```

Comment:
You can unify the concept of this `piece`, like `slot`? and change `payment_by_instalments` to `slot_pay`? This will be understandable.